core/peerset/src/peersstate.rs

```diff
@@ -0,0 +1,529 @@
+				reserved && !connection_state.is_connected()
+			})
+			.map(|(peer_id, _)| peer_id.clone());
+
+		if let Some(peer_id) = peer_id {
+			Some(NotConnectedPeer {
+				parent: self,
+				peer_id: Cow::Owned(peer_id),
+			})
+		} else {
+			None
+		}
+	}
+
+	/// Returns the peer with the highest reputation and that we are not connected to.
+	///
+	/// If multiple nodes have the same reputation, which one is returned is unspecified.
+	pub fn highest_not_connected_peer(&mut self) -> Option<NotConnectedPeer> {
+		let peer_id = self.nodes
+			.iter()
+			.filter(|(_, Node { connection_state, .. })| !connection_state.is_connected())
+			.fold(None::<(&PeerId, &Node)>, |mut cur_node, to_try| {
+				if let Some(cur_node) = cur_node.take() {
+					if cur_node.1.reputation >= to_try.1.reputation {
+						return Some(cur_node);
+					}
+				}
+				Some(to_try)
+			})
+			.map(|(peer_id, _)| peer_id.clone());
+
+		if let Some(peer_id) = peer_id {
+			Some(NotConnectedPeer {
+				parent: self,
+				peer_id: Cow::Owned(peer_id),
+			})
+		} else {
+			None
+		}
+	}
+}
+
+/// Grants access to the state of a peer in the `PeersState`.
+pub enum Peer<'a> {
+	/// We are connected to this node.
+	Connected(ConnectedPeer<'a>),
+	/// We are not connected to this node.
+	NotConnected(NotConnectedPeer<'a>),
+	/// We have never heard of this node.
+	Unknown(UnknownPeer<'a>),
+}
+
+impl<'a> Peer<'a> {
+	/// If we are the `Connected` variant, returns the inner `ConnectedPeer`. Returns `None`
+	/// otherwise.
+	pub fn into_connected(self) -> Option<ConnectedPeer<'a>> {
+		match self {
+			Peer::Connected(peer) => Some(peer),
+			Peer::NotConnected(_) => None,
+			Peer::Unknown(_) => None,
+		}
+	}
+
+	/// If we are the `Unknown` variant, returns the inner `ConnectedPeer`. Returns `None`
+	/// otherwise.
+	#[cfg(test)]	// Feel free to remove this if this function is needed outside of tests
+	pub fn into_not_connected(self) -> Option<NotConnectedPeer<'a>> {
+		match self {
+			Peer::Connected(_) => None,
+			Peer::NotConnected(peer) => Some(peer),
+			Peer::Unknown(_) => None,
+		}
+	}
+
+	/// If we are the `Unknown` variant, returns the inner `ConnectedPeer`. Returns `None`
+	/// otherwise.
+	#[cfg(test)]	// Feel free to remove this if this function is needed outside of tests
+	pub fn into_unknown(self) -> Option<UnknownPeer<'a>> {
+		match self {
+			Peer::Connected(_) => None,
+			Peer::NotConnected(_) => None,
+			Peer::Unknown(peer) => Some(peer),
+		}
+	}
+}
+
+/// A peer that is connected to us.
+pub struct ConnectedPeer<'a> {
+	parent: &'a mut PeersState,
+	peer_id: Cow<'a, PeerId>,
+}
+
+impl<'a> ConnectedPeer<'a> {
+	/// Destroys this `ConnectedPeer` and returns the `PeerId` inside of it.
+	pub fn into_peer_id(self) -> PeerId {
+		self.peer_id.into_owned()
+	}
+
+	fn state(&self) -> &Node {
+		self.parent.nodes.get(&self.peer_id)
+			.expect("We only ever build a ConnectedPeer if the node's in the list; QED")
```

Comment:
> We only ever build a ConnectedPeer if the node's in the list; QED 

This is not a strong proof.

If the node is in what list? At what point would it need to be? What stops the list mutating after the `ConnectedPeer` has been built but before `state` is called?

Reply:
The `PeersState` struct is essentially a glorified `HashMap`, and a `ConnectedPeer` is the equivalent of `OccupiedEntry` (https://doc.rust-lang.org/nightly/std/collections/hash_map/struct.OccupiedEntry.html).

The contract of `ConnectedPeer` is "as long as this object is alive, we guarantee that the peer is in list and in the connected state". It holds a mutable borrows the `PeersState`, meaning that nothing else can change it.

I don't really know how to provide a brief explanation of that, though.

Reply:
I think I'll change it to not have any possible panic instead.

Reply:
this doesn't sound like a reasonable proof.

proofs cannot be of the form "the program state is valid; this thing always succeeds when the program state is valid; qed".

this is because they cannot presume continuous-correct-state of the program (at least not unless there's a completely trivial means of auditing it).

Reply:
I pushed a new commit that expands a bit.

> at least not unless there's a completely trivial means of auditing it

It is extremely simple to prove that the code here is always in a valid state. Nothing is racy, everything is single-threaded, everything mutably borrows.

The two states that have to be in sync (the presence of a node in a list and the fact that `ConnectedPeer`/`NotConnectedPeer` is alive) only concerns 2 locations in the code, and it is done in a very straight-forward way.

If this still isn't ok, I'll do the change that removes any possible panic.

Reply:
a small overhead is fine. proofs should not extend beyond the scope that they're in.